tools/mason/MasonModify.chpl

```diff
@@ -0,0 +1,267 @@
+    toml["dependencies"] = deps;
+    toml["dependencies"][toAdd] = version;
+  }
+  return toml;
+}
+
+/* Remove a mason dependency from Mason.toml */
+private proc masonRemove(toml: unmanaged Toml, toRm: string) throws {
+  if toml.pathExists("dependencies") {
+    if toml.pathExists("dependencies." + toRm) {
+      toml["dependencies"].D.remove(toRm);
+    }
+    else {
+      throw new MasonError("No dependency exists by that name");
+    }
+  }
+  else {
+    throw new MasonError("No dependencies");
+  }
+  return toml;
+}
+
+/* Add a system dependency to Mason.toml */
+private proc masonSystemAdd(toml: unmanaged Toml, toAdd: string, version: string) throws {
+  
+  if toml.pathExists("system") {
+    if toml.pathExists("system." + toAdd) {
+      throw new MasonError("A dependency by that name already exists in Mason.toml");
+    }
+    else {
+      toml["system"][toAdd] = version;
+    }
+  }
+  else {
+    var pkgdom: domain(string);
+    var pkgdeps: [pkgdom] unmanaged Toml;
+    toml["system"] = pkgdeps;
+    toml["system"][toAdd] = version;
+  }
+  return toml;
+}
+
+/* Remove a system dependency from Mason.toml */
+private proc masonSystemRemove(toml: unmanaged Toml, toRm: string) throws {
+  if toml.pathExists("system") {
+    if toml.pathExists("system." + toRm) {
+      toml["system"].D.remove(toRm);
+    }
+    else {
+      throw new MasonError("No system dependency exists by " + toRm);
+    }
+  }
+  else {
+    throw new MasonError("No system dependency exists by " + toRm);
+  }
+  return toml;
+}
+
+/* Add an external dependency to Mason.toml */
+private proc masonExternalAdd(toml: unmanaged Toml, toAdd: string, version: string) throws {
+  if toml.pathExists("external") {
+    if toml.pathExists("external." + toAdd) {
+      throw new MasonError("An external dependency by that name already exists in Mason.toml");
+    }
+    else {
+      toml["external"][toAdd] = version;
+    }
+  }
+  else {
+    var exdom: domain(string);
+    var exdeps: [exdom] unmanaged Toml;
+    toml["external"] = exdeps;
+    toml["external"][toAdd] = version;
+  }
+  return toml;
+}
+
+/* Remove an external dependency from Mason.toml */
+private proc masonExternalRemove(toml: unmanaged Toml, toRm: string) throws {
+  if toml.pathExists("external") {
+    if toml.pathExists("external." + toRm) {
+      toml["external"].D.remove(toRm);
+    }
+    else {
+      throw new MasonError("No external dependency exists by that name");
+    }
+  }
+  else {
+    throw new MasonError("No external dependency exists by that name");
+  }
+  return toml;
+}
+
+/* Generate the modified Mason.toml */
+private proc generateToml(toml: Toml, tomlPath: string) {
+  const tomlFile = open(tomlPath, iomode.cw);
+  const tomlWriter = tomlFile.writer();
+  tomlWriter.writeln(toml);
+  tomlWriter.close();
+  tomlFile.close();
+}
+
+private proc checkVersion(version: string) throws {
+  const pattern = compile("([0-9].[0-9].[0-9][a-zA-Z]?)");
```

Comment:
What is the purpose of the `[a-zA-Z]`?

Reply:
Some packages include letters in their version numbers. For instance, the current version of `openSSL` that is installed on my mac is `0.9.8zh`.

Reply:
Gotcha, so this is a general case regex (handling normal, system, and external packages).